src/pmdas/amdgpu/amdgpu.c

```diff
@@ -0,0 +1,547 @@
+  unsigned int item = pmID_item(mdesc->m_desc.pmid);
+
+  if (item != 0 && cluster == 0 && inst > indomtab[GCARD_INDOM].it_numinst)
+    return PM_ERR_INST;
+
+  switch (cluster) {
+  case 0: /* amdgpu general and per-card metrics */
+    switch (item) {
+    case AMDGPU_NUMCARDS:
+      atom->ul = pcp_amdgpuinfo.numcards;
+      break;
+    case AMDGPU_SAMPLES:
+      atom->ull = pcp_amdgpuinfo.info[inst].samples;
+      break;
+    case AMDGPU_CARDNAME:
+      if (pcp_amdgpuinfo.info[inst].failed[AMDGPU_CARDNAME])
+        return PM_ERR_VALUE;
+      atom->cp = pcp_amdgpuinfo.info[inst].name;
+      break;
+    case AMDGPU_MEMUSED:
+      if (pcp_amdgpuinfo.info[inst].failed[AMDGPU_MEMUSED])
+        return PM_ERR_VALUE;
+      atom->ull = pcp_amdgpuinfo.info[inst].memory.used;
+      break;
+    case AMDGPU_MEMTOTAL:
+      if (pcp_amdgpuinfo.info[inst].failed[AMDGPU_MEMTOTAL])
+        return PM_ERR_VALUE;
+      atom->ull = pcp_amdgpuinfo.info[inst].memory.total;
+      break;
+    case AMDGPU_MEMFREE:
+      if (pcp_amdgpuinfo.info[inst].failed[AMDGPU_MEMFREE])
+        return PM_ERR_VALUE;
+      atom->ull =
+	pcp_amdgpuinfo.info[inst].memory.usable -
+	pcp_amdgpuinfo.info[inst].memory.used;
+      break;
+    case AMDGPU_MEMUSED_ACCUM:
+      if (pcp_amdgpuinfo.info[inst].failed[AMDGPU_MEMUSED_ACCUM])
+        return PM_ERR_VALUE;
+      atom->ull = pcp_amdgpuinfo.info[inst].memaccum;
+      break;
+    case AMDGPU_GPU_CLOCK:
+      if (pcp_amdgpuinfo.info[inst].failed[AMDGPU_GPU_CLOCK])
+        return PM_ERR_VALUE;
+      /* The GPU speed is the memory clock (GFX_MCLK)
+       * The GDDRx memory speed is the shader clock (GFX_SCLK)
+       * In MHz
+       */
+      atom->ul = pcp_amdgpuinfo.info[inst].gpu_clock;
+      break;
+    case AMDGPU_GPU_CLOCK_MAX:
+      if (pcp_amdgpuinfo.info[inst].failed[AMDGPU_GPU_CLOCK_MAX])
+        return PM_ERR_VALUE;
+      /* The GPU max speed is the max_memory_clk.
+       * The GDDRx memory max speed is max_engine_clk
+       * In MHz
+       */
+      atom->ul = pcp_amdgpuinfo.info[inst].gpu_info.max_memory_clk;
+      break;
+    case AMDGPU_MEMORY_CLOCK:
+      if (pcp_amdgpuinfo.info[inst].failed[AMDGPU_MEMORY_CLOCK])
+        return PM_ERR_VALUE;
+      /* The GPU speed is the memory clock (GFX_MCLK)
+       * The GDDRx memory speed is the shader clock (GFX_SCLK)
+       * In MHz
+       */
+      atom->ul = pcp_amdgpuinfo.info[inst].mem_clock;
+      break;
+    case AMDGPU_MEMORY_CLOCK_MAX:
+      if (pcp_amdgpuinfo.info[inst].failed[AMDGPU_MEMORY_CLOCK_MAX])
+        return PM_ERR_VALUE;
+      /* The GPU max speed is the max_memory_clk.
+       * The GDDRx memory max speed is max_engine_clk
+       * In MHz
+       */
+      atom->ul = pcp_amdgpuinfo.info[inst].gpu_info.max_engine_clk;
+      break;
+    case AMDGPU_TEMPERATURE:
+      if (pcp_amdgpuinfo.info[inst].failed[AMDGPU_TEMPERATURE])
+        return PM_ERR_VALUE;
+      /* In millidegrees Celsius */
```

Comment:
One possibility here is to add a label to this metric describing the temperature units (since temperature this isn't a typical performance measure captured in regular PCP metrics).  src/pmdas/denki/denki.c has some examples of custom units labels for various power measures, for reference.

Reply:
I see the label callbacks now but should we add "units":"millidegrees celcius" as a label for this one?